sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java

```diff
@@ -272,52 +338,315 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
+      @FinishBundle
+      public void finishBundle() throws Exception {
+        checkForFailures();
+
+        // Flush all outstanding records, blocking call
+        flushAll();
+      }
+
+      @Teardown
+      public void tearDown() throws Exception {
+        if (producer != null) {
+          producer.destroy();
+          producer = null;
+        }
+      }
+
+      /**
+       * Flush outstanding records until the total number will be less than required or
+       * number of retries will be exhausted.
+       */
+      private void flush(int numMax) throws InterruptedException, IOException {
+        int retries = spec.getRetries();
+        int numOutstandingRecords = producer.getOutstandingRecordsCount();
+
+        while (numOutstandingRecords > numMax && retries-- > 0) {
+          producer.flush();
+          // wait until outstanding records will be flushed
+          Thread.sleep(1000);
+          numOutstandingRecords = producer.getOutstandingRecordsCount();
+        }
+
+        if (numOutstandingRecords > numMax) {
+          String message =
+              String.format(
+                  "Number of outstanding records [%d] are greater than required [%d].",
+                  numOutstandingRecords,
+                  numMax);
+          LOG.error(message);
+          throw new IOException(message);
+        }
+      }
+
+      private void flushAll() throws InterruptedException, IOException {
+        flush(0);
+      }
+
+      /**
+       * If any write has asynchronously failed, fail the bundle with a useful error.
+       */
+      private void checkForFailures() throws IOException {
+        // Note that this function is never called by multiple threads and is the only place that
+        // we remove from failures, so this code is safe.
+        if (failures.isEmpty()) {
+          return;
+        }
+
+        StringBuilder logEntry = new StringBuilder();
+        int i = 0;
+        while (!failures.isEmpty()) {
+          i++;
+          KinesisWriteException exc = failures.remove();
+
+          logEntry.append("\n").append(exc.getMessage());
+          Throwable cause = exc.getCause();
+          if (cause != null) {
+            logEntry.append(": ").append(cause.getMessage());
+
+            if (cause instanceof UserRecordFailedException) {
+              List<Attempt> attempts = ((UserRecordFailedException) cause).getResult()
+                  .getAttempts();
+              for (Attempt attempt : attempts) {
+                if (attempt.getErrorMessage() != null) {
+                  logEntry.append("\n").append(attempt.getErrorMessage());
+                }
+              }
+            }
+          }
+        }
+        failures.clear();
+
+        String message =
+            String.format(
+                "Some errors occurred writing to Kinesis. First %d errors: %s",
+                i,
+                logEntry.toString());
+        LOG.error(message);
```

Comment:
It's not necessary to both log and throw an error - an exception will surely be logged at a higher level.

Reply:
Ok, agree.